graphql_compiler/compiler/ir_lowering_cypher/ir_lowering.py

```diff
@@ -0,0 +1,184 @@
+    location_translations = make_revisit_location_translations(query_metadata_table)
+    visitor_fn = make_location_rewriter_visitor_fn(location_translations)
+
+    new_ir_blocks = []
+    for block in ir_blocks:
+        if isinstance(block, MarkLocation) and block.location in location_translations:
+            # Drop this block, since we'll be replacing its location with its revisit origin.
+            pass
+        else:
+            # Rewrite the locations in this block (if any), to reflect the desired translation.
+            new_block = block.visit_and_update_expressions(visitor_fn)
+            new_ir_blocks.append(new_block)
+
+    return new_ir_blocks
+
+
+def replace_local_fields_with_context_fields(ir_blocks):
+    """Rewrite LocalField expressions into ContextField expressions referencing that location."""
+    def visitor_func_base(location, expression):
+        """Rewriter function that converts LocalFields into ContextFields at the given location."""
+        if not isinstance(expression, LocalField):
+            return expression
+
+        location_at_field = location.navigate_to_field(expression.field_name)
+        if isinstance(location, FoldScopeLocation):
+            return FoldedContextField(location_at_field, expression.field_type)
+        else:
+            return ContextField(location_at_field, expression.field_type)
+
+    new_ir_blocks = []
+    blocks_to_be_rewritten = []
+    for block in ir_blocks:
+        if isinstance(block, MarkLocation):
+            # First, rewrite all the blocks that might have referenced this location.
+            visitor_fn = partial(visitor_func_base, block.location)
+            for block_for_rewriting in blocks_to_be_rewritten:
+                new_block = block_for_rewriting.visit_and_update_expressions(visitor_fn)
+                new_ir_blocks.append(new_block)
+
+            # Then, append the MarkLocation block itself and start with an empty rewrite list.
+            blocks_to_be_rewritten = []
+            new_ir_blocks.append(block)
+        else:
+            blocks_to_be_rewritten.append(block)
+
+    # Append any remaining blocks that did not need rewriting.
+    new_ir_blocks.extend(blocks_to_be_rewritten)
+
+    if len(ir_blocks) != len(new_ir_blocks):
+        raise AssertionError(u'The number of IR blocks unexpectedly changed, {} vs {}: {} {}'
+                             .format(len(ir_blocks), len(new_ir_blocks), ir_blocks, new_ir_blocks))
+
+    return new_ir_blocks
+
+
+def move_filters_in_optional_locations_to_global_operations(cypher_query, query_metadata_table):
+    """Move Filter blocks found within @optional traversals to the global operations section.
+
+    This transformation is necessary to uphold the compiler's chosen semantics around filters
+    within optional traversals: if the edge exists but the filter fails to match, we can't pretend
+    the edge didn't exist. The Cypher specification chooses the opposite approach. The
```

Comment:
Can you elaborate a bit more what you mean by "if the edge exists but the filter fails to match, we can't pretend
the edge didn't exist"?

Reply:
Suppose we have two Person nodes with names Albert and Betty such that there's an edge Albert-knows->Betty. 

Then we run a query

```
MATCH (p:Person {name:"Albert"})
OPTIONAL MATCH (p)-[r:KNOWS]->(p2:Person)
where p2.name="Charles"
return p,r,p2
```

Neo4j returns `Albert` for `p`, `null` for `r`, and null for `p2`.

In this case, we do have an "knows" edge from Albert to some other person but the filter fails because the other person isn't named Charles. However, if Albert has no outgoing "knows" edges, then the filter wouldn't apply at all, so effectively Neo4j pretends that the outgoing edge doesn't exist.

Reply:
Why can't we pretend that the edge does not exist? Isn't that exactly what we are supposed to do?

Reply:
Per our spec: https://github.com/kensho-technologies/graphql-compiler#constraints-and-rules

```When filtering (via @filter) or type coercion (via e.g. ... on Animal) are applied at or within a vertex field marked @optional, the @optional is given precedence:

- If a given result set cannot produce a value for the optional vertex field, it is preserved: the @optional directive is applied first, and no filtering or type coercion can happen.
- If a given result set is able to produce a value for the optional vertex field, the @optional does not apply, and that value is then checked against the filtering or type coercion. These subsequent operations may then cause the result set to be discarded if it does not match.```

So we may not pretend that the edge does not exist.

Reply:
Btw, @LWprogramming the `Albert-knows->Betty` example is really good. Consider adding (maybe a query-language-agnostic variant) to the spec section that I quoted? I think it might really help give people some intuition about what is going on there.

Reply:
Sure! I'll make a PR.